lib/web/terminal.go

```diff
@@ -221,6 +222,31 @@ func (t *TerminalHandler) Close() error {
 	return nil
 }
 
+// startPingLoop starts a loop that will continuously send a ping frame through the websocket
+// to prevent the connection between web client and teleport proxy from becoming idle.
+// Interval is determined by the keep_alive_interval config set by user (or default).
+// Loop will terminate when there is an error sending ping frame or when terminal session is closed.
+func (t *TerminalHandler) startPingLoop(ws *websocket.Conn) {
+	t.log.Debugf("Starting websocket ping loop with interval %v.", t.params.KeepAliveInterval)
+	tickerCh := time.NewTicker(t.params.KeepAliveInterval)
+	defer tickerCh.Stop()
+
+	for {
+		select {
+		case <-tickerCh.C:
+			deadline := time.Now().Add(t.params.KeepAliveInterval)
+			if err := ws.WriteControl(websocket.PingMessage, nil, deadline); err != nil {
+				t.log.Errorf("Unable to send ping frame to web client: %v.", err)
```

Comment:
Suppose there is a temporary availability issue, should we stop the ping loop then? Won't that cause more problems?

Reply:
Well this stops the web terminal on error too which is what the pre-v9 implementation does. I have added a case that checks if the error is temporary as per the net package and retries the ping frame if it is.